app/models/pipeline_run_stage.rb

```diff
@@ -49,12 +59,35 @@ def started?
     job_command.present?
   end
 
-  def failed?
-    job_status == STATUS_FAILED
+  def stage_status_file(status)
+    basename = "#{job_id}.#{status}"
+    s3_folder = case name
+                when HOST_FILTERING_STAGE_NAME
+                  pipeline_run.host_filter_output_s3_path
+                when ALIGNMENT_STAGE_NAME
+                  pipeline_run.alignment_output_s3_path
+                when POSTPROCESS_STAGE_NAME
+                  pipeline_run.postprocess_output_s3_path
+                end
+    "#{s3_folder}/#{basename}"
+  end
+
+  def assess_and_update(status_file_suffix, job_status_value)
```

Comment:
In other words, you check for file existence, and cache that in the SQL DB?

Reply:
Yes. It assesses `succeeded?` or `failed?` by looking at a file in S3. No need to query S3 if we have already done it before. That was one of the problems that caused the project page to slow down -- `fetch_pipeline_version` always had to make a request to S3 from scratch because it never stored the result it had just found.